src/map/itemdb.cpp

```diff
@@ -2274,194 +2272,396 @@ bool itemdb_isNoEquip(struct item_data *id, uint16 m) {
-	return ((struct s_random_opt_data*)uidb_get(itemdb_randomopt, id));
+const std::string RandomOptionDatabase::getDefaultLocation() {
+	return std::string(db_path) + "/item_randomopt_db.yml";
 }
 
-/** Random option
-* <ID>,<{Script}>
-*/
-static bool itemdb_read_randomopt(const char* basedir, bool silent) {
-	uint32 lines = 0, count = 0;
-	char line[1024];
+/**
+ * Reads and parses an entry from the item_randomopt_db.
+ * @param node: YAML node containing the entry.
+ * @return count of successfully parsed rows
+ */
+uint64 RandomOptionDatabase::parseBodyNode(const YAML::Node &node) {
+	uint16 id;
 
-	char path[256];
-	FILE* fp;
+	if (!this->asUInt16(node, "Id", id))
+		return 0;
 
-	sprintf(path, "%s/%s", basedir, "item_randomopt_db.txt");
+	std::shared_ptr<s_random_opt_data> randopt = this->find(id);
+	bool exists = randopt != nullptr;
 
-	if ((fp = fopen(path, "r")) == NULL) {
-		if (silent == 0) ShowError("itemdb_read_randomopt: File not found \"%s\".\n", path);
-		return false;
+	if (!exists) {
+		if (!this->nodesExist(node, { "Option", "Script" }))
+			return 0;
+
+		randopt = std::make_shared<s_random_opt_data>();
+		randopt->id = id;
 	}
 
-	while (fgets(line, sizeof(line), fp)) {
-		char *str[2], *p;
+	if (this->nodeExists(node, "Option")) {
+		std::string name;
 
-		lines++;
+		if (!this->asString(node, "Option", name))
+			return 0;
 
-		if (line[0] == '/' && line[1] == '/') // Ignore comments
-			continue;
+		if (random_option_db.option_search_name(name)) {
+			this->invalidWarning(node["Option"], "Found duplicate random option name for %s, skipping.\n", name.c_str());
+			return 0;
+		}
 
-		memset(str, 0, sizeof(str));
+		randopt->name = name;
+	}
 
-		p = line;
+	if (this->nodeExists(node, "Script")) {
+		std::string script;
 
-		p = trim(p);
+		if (!this->asString(node, "Script", script))
+			return 0;
 
-		if (*p == '\0')
-			continue;// empty line
+		if (randopt->script) {
+			aFree(randopt->script);
+			randopt->script = nullptr;
+		}
 
-		if (!strchr(p, ','))
-		{
-			ShowError("itemdb_read_randomopt: Insufficient columns in line %d of \"%s\", skipping.\n", lines, path);
-			continue;
+		randopt->script = parse_script(script.c_str(), this->getCurrentFile().c_str(), id, SCRIPT_IGNORE_EXTERNAL_BRACKETS);
+	}
+
+	if (!exists)
+		this->put(id, randopt);
+
+	return 1;
+}
+
+RandomOptionDatabase random_option_db;
+
+/**
+ * Check if the given random option name exists.
+ * @param name: Random option name
+ * @return True on success or false on failure
+ */
+bool RandomOptionDatabase::option_search_name(std::string name) {
+	for (const auto &opt : random_option_db) {
+		if (opt.second->name.compare(name) == 0)
+			return true;
+	}
+
+	return false;
+}
+
+/**
+ * Return the constant value of the given random option.
+ * @param name: Random option name
+ * @param id: Random option ID
+ * @return True on success or false on failure
+ */
+bool RandomOptionDatabase::option_get_id(std::string name, uint16 *id) {
```

Comment:
Make id a reference, this way it cannot be nullptr and assignment does not require dereferencing.